topics/RDMA.md

```diff
@@ -0,0 +1,170 @@
+RDMA. To test a server supports RDMA or not:
+
+    ~# rdma dev show (a new version iproute2 package)
+Or:
+
+    ~# ibv_devices (ibverbs-utils package of Debian/Ubuntu)
+
+
+## Protocol
+
+The protocol defines the QP type RC (like TCP), communication commands, and payload exchange mechanism.
+This dependency is based solely on the RDMA (aka Infiniband) specification
+and is independent of both software (including the OS and user libraries)
+and hardware (including vendors and low-level transports).
+
+Valkey Over RDMA has control-plane (control messages) and data-plane (payload transfer).
+
+### Control message
+
+Control messages use fixed 32-byte big-endian message structures:
+```C
+typedef struct ValkeyRdmaFeature {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    /* select features */
+    uint16_t select;
+    uint8_t rsvd[20];
+    /* feature bits */
+    uint64_t features;
+} ValkeyRdmaFeature;
+
+typedef struct ValkeyRdmaKeepalive {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[30];
+} ValkeyRdmaKeepalive;
+
+typedef struct ValkeyRdmaMemory {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[14];
+    /* address of a transfer buffer which is used to receive remote streaming data,
+     * aka 'RX buffer address'. The remote side should use this as 'TX buffer address' */
+    uint64_t addr;
+    /* length of the 'RX buffer' */
+    uint32_t length;
+    /* the RDMA remote key of 'RX buffer' */
+    uint32_t key;
+} ValkeyRdmaMemory;
+
+typedef union ValkeyRdmaCmd {
+    ValkeyRdmaFeature feature;
+    ValkeyRdmaKeepalive keepalive;
+    ValkeyRdmaMemory memory;
+} ValkeyRdmaCmd;
+```
+
+### Opcodes
+|Command| Value | Description |
+| :----: | :----: | :----: |
+| `GetServerFeature`   | 0 | required, get the features offered by Valkey server |
+| `SetClientFeature`   | 1 | required, negotiate features and set it to Valkey server |
+| `Keepalive`          | 2 | required, detect unexpected orphan connection |
+| `RegisterXferMemory` | 3 | required, tell the 'RX transfer buffer' information to the remote side, and the remote side uses this as 'TX transfer buffer' |
+
+### RDMA Operations
+- Send a control message by RDMA '**`ibv_post_send`**' with opcode '**`IBV_WR_SEND`**' with structure
+  'ValkeyRdmaCmd'.
+- Receive a control message by RDMA '**`ibv_post_recv`**', and the received buffer
+  size should be size of 'ValkeyRdmaCmd'.
+- Transfer stream data by RDMA '**`ibv_post_send`**' with opcode '**`IBV_WR_RDMA_WRITE`**' (optional) and
+  '**`IBV_WR_RDMA_WRITE_WITH_IMM`**' (required), to write data segments into a connection by
+  RDMA [WRITE][WRITE][WRITE]...[WRITE WITH IMM], the length of total buffer is described by
+  immediate data (unsigned int 32).
+
+
+### Maximum WQEs of RDMA
+o specific limit, 1024 recommended for WQEs. Recommended WQEs is 1024.
+Flow control for WQE MAY be defined/implemented in the future.
```

Comment:
We say 1024 is recommended twice, but I don't know what this is or why it matters. Do we need this sentence? Would someone implementing a client know what to do with this?

Reply:
What about `No specific limit, 1024 recommended for WQEs. Flow control for WQE MAY be defined/implemented in the future.`

For RDMA, before receiving any message, the QP must prepare the receive buffer by `ibv_post_recv`, otherwise the incoming message can't find any available WQE of receive queue, then error occurs.

When Valkey client gets connected, `command` and `info` commands will trigger lots of message returned, to avoid no enough WQEs issue, 1024 is recommended.